text/0000-cfg-path-version.md

```diff
@@ -0,0 +1,674 @@
+it even once unless the name of the path changes in-between.
+
+### Preventing relative paths
+
+The reason why we have enforced that all paths must start with `::` inside
+`accessible(..)` is that if we allow relative paths, and users write
+`accessible(self::foo)`, then they can construct situations such as:
+
+```
+#[cfg(accessible(self::bar)]
+fn foo() {}
+
+#[cfg(accessible(self::foo)]
+fn bar() {}
+```
+
+One way around this is to collect all items before `cfg`-stripping,
+but this can cause problems with respect to stage separation.
+Therefore, we prevent this from occurring with a simple syntactic check.
+
+### `#[cfg(accessible(..))` or `#[cfg(accessible = ..)`
+
+We need to decide between the syntax `accessible(..)` or `accessible = ..`.
+The reason we've opted for the former rather than the latter is that the
+former syntax looks more like a question/query whilst the latter looks more
+like a statement of fact.
+
+In addition, if we would like to enable `accessible = $path` we would need to
+extend the meta grammar. We could justify that change in and of itself by
+observing that crates such as `serde_derive` permit users to write things like
+`#[serde(default = "some::function")]`. By changing the grammar we can allow
+users to instead write: `#[serde(default = some::function)]`.
+However, in this case, `accessible($path)` seems the optimal notation.
+
+If we would like to extend the meta grammar, we could do so by changing:
+
+```
+named_value : "=" lit ;
+
+meta_or_lit : meta | lit ;
+meta_or_lit_list : meta_or_lit "," meta_or_lit_list ","? ;
+meta_list : "(" meta_or_lit_list ")" ;
+meta : path ( named_value | meta_list )? ;
+```
+
+into:
+
+```
+lit_or_path : path | lit ;
+named_value : "=" lit_or_path ;
+
+meta_or_lit : meta | lit ;
+meta_or_lit_list : meta_or_lit "," meta_or_lit_list ","? ;
+meta_list : "(" meta_or_lit_list ")" ;
+meta : path ( named_value | meta_list )? ;
+```
+
+### The bikeshed
+
+One might consider other names for the flag instead of `accessible`.
+Some contenders are:
+
++ `path_accessible`
++ `can_use`
++ `path_exists`
++ `have_path`
++ `have`
++ `have_item`
++ `path_reachable`
++ `item_reachable`
++ `item_exists`
+
+Currently `accessible` is the choice because it clearly signals the intent
+while also being short enough to remain ergonomic to use.
+In particular, while `path_accessible` might be somewhat more unambiguous,
+we argue that from the context of seeing `accessible(::std::foo::bar)`
+it is clear that it is paths we are talking about because the argument
+is a path and not something else.
+
+While `can_use` is also a strong contender, we reject this option because
+it may imply to the user that only things that you may `use $path;` can
+go in there. Meanwhile, you may `#[cfg(accessible(::foo::MyTrait::my_method))`
+which is *not* possible as `use ::foo::MyTrait::my_method;`. This also applies
+to other associated items and inherent methods as well as `struct` fields.
```

Comment:
I agree somewhat, but I still like `can_use`. I'm fine with `accessible(path)`, too, I guess.
(I also just realized `usable(path)` could be added alongside `can_use`, as possible alternatives)

Reply:
Added to the list ;)